curator-client/src/main/java/org/apache/curator/ConnectionState.java

```diff
@@ -160,13 +162,33 @@ public void process(WatchedEvent event)
             }
         }
 
+        // only wait during tests
+        assert waitOnExpiredEvent(event.getState());
```

Comment:
I don't think that it's reasonable to assume that assertions will only be turned on during testing. If you look at something like LeaderSelector, it has specific code in there to support unit testing (the debugLeadershipLatch variable).

If you can cause the problem to occur without this code though, then it should be removed. I had a quick play with it and I couldn't seem to reproduce it without this code though.